[AArch64][SVE2] Generate urshr rounding shift rights

UsmanNadeem opened the pull request:
Matching code is taken from rshrnb except that immediate
shift value has a larger range, and support for signed shift. rshrnb code
now checks for the new AArch64ISD::URSHR_I_PRED node to convert
to narrowing rounding op.

Please let me know if there is a better way/location to do this!

UsmanNadeem committed:
[AArch64][SVE2] Generate signed/unsigned rounding shift rights

Matching code is similar to that for rshrnb except that immediate
shift value has a larger range, and support for signed shift. rshrnb
now uses the new AArch64ISD node for uniform rounding.

Change-Id: Idbb811f318d33c7637371cf7bb00285d20e1771d

davemgreen reviewed (commented):
Hello. Some of these instructions have some awkward edge cases. Is this correct if the add overflows?

UsmanNadeem commented:
> Hello. Some of these instructions have some awkward edge cases. Is this correct if the add overflows?

Yes, the result will be correct. Meaning that if we operate on the max value for the respective data type, the result will not be effected by an overflow.
e.g. for uint8:
`255 r/ 4 == (255 + (1<<2)) >> 2  == urshr z0.b, p0/m, z0.b, #2` the result will be `64`.

davemgreen commented:
I agree that is what urshr will produce but not the shr+add. The add would overflow, producing zero, so the shifted value is zero.
I think it needs to be something like trunc(shr(add(ext))) or maybe we check that at least one of the top bits is known to not be 1 (so the add doesn't overflow).

UsmanNadeem commented:
> I agree that is what urshr will produce but not the shr+add. The add would overflow, producing zero, so the shifted value is zero. I think it needs to be something like trunc(shr(add(ext))) or maybe we check that at least one of the top bits is known to not be 1 (so the add doesn't overflow).

Right, I guess in this case the existing code for rshrnb also has this bug. I'll work on it.

UsmanNadeem committed:
Handle wide operations, Account for overflows, remove signed shift

Change-Id: I7450629fa43bb3ac1bc40daaa760255eed483c10

UsmanNadeem commented:
- Handled overflows.
- Added the following combine for wide shifts: `uzp1(urshr(uunpklo(X),C), urshr(uunpkhi(X), C)) -> urshr(X, C)`
- Removed code for signed RSHR for now.

davemgreen reviewed (commented):
Using nuw sounds like a smart idea, and should handle cases like https://godbolt.org/z/rf1oaMvd4 as the nuw get automatically added. (It does make some of the tools I have for trying to analyze equivalence less useful, but as far as I can tell it seems OK). There might be an extension too where it checks the demanded bits would make the overflow unimportant, but I'm not sure where to make that happen.

UsmanNadeem committed:
fix overflow handling

Change-Id: Id6dceead02c7473ed5c3635c2b56c7f367315563

UsmanNadeem commented on llvm/test/CodeGen/AArch64/sve2-intrinsics-combine-rshrnb.ll:
```diff
@@ -10,7 +10,7 @@ define void @add_lshr_rshrnb_b_6(ptr %ptr, ptr %dst, i64 %index){
 ; CHECK-NEXT:    ret
   %load = load <vscale x 8 x i16>, ptr %ptr, align 2
-  %1 = add <vscale x 8 x i16> %load, trunc (<vscale x 8 x i32> shufflevector (<vscale x 8 x i32> insertelement (<vscale x 8 x i32> poison, i32 32, i64 0), <vscale x 8 x i32> poison, <vscale x 8 x i32> zeroinitializer) to <vscale x 8 x i16>)
+  %1 = add nuw <vscale x 8 x i16> %load, trunc (<vscale x 8 x i32> shufflevector (<vscale x 8 x i32> insertelement (<vscale x 8 x i32> poison, i32 32, i64 0), <vscale x 8 x i32> poison, <vscale x 8 x i32> zeroinitializer) to <vscale x 8 x i16>)
```
I think you are correct. I have fixed it.

UsmanNadeem commented:
- `canLowerSRLToRoundingShiftForVT` returns true if the add has NUW or if the number of bits used in the return value allow us to not care about the overflow (tested by rshrnb).

- `srl(add(X, 1 << (ShiftValue - 1)), ShiftValue)` is transformed to `urshr` or `rshrnb` if the result it truncated.

- `uzp1(rshrnb(uunpklo(X),C), rshrnb(uunpkhi(X), C))` is converted to `urshr(X, C)`, tested by the wide_trunc tests.

davemgreen commented on llvm/test/CodeGen/AArch64/sve2-rsh.ll:
```diff
@@ -0,0 +1,290 @@
+; CHECK-NEXT:    ret
+  %add = add nuw nsw <vscale x 2 x i64> %x, splat (i64 32)
+  %sh = lshr <vscale x 2 x i64> %add, splat (i64 6)
+  call void @use(<vscale x 2 x i64> %add)
```
Can you store the value to introduce an extra use.

UsmanNadeem committed:
address comments

Change-Id: I076f19c947696100ec469c8407b6d235d6444145

UsmanNadeem commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20895,12 +20983,24 @@ static SDValue performUzpCombine(SDNode *N, SelectionDAG &DAG,
 
+  // uzp1(bitcast(x), bitcast(y)) -> uzp1(x, y)
+  if (isHalvingTruncateAndConcatOfLegalIntScalableType(N) &&
+      Op0.getOpcode() == ISD::BITCAST && Op1.getOpcode() == ISD::BITCAST) {
```
This is the transform:
```
    nxv4i32 = AArch64ISD::UZP1 bitcast(nxv4i32 to nxv2i64), bitcast(nxv4i32 to nxv2i64)
    =>
    nxv4i32 = AArch64ISD::UZP1 nxv2i64, nxv2i64
```

Both get lowered to uzp1.s and removing the bitcast makes the code simpler.

An example here: https://godbolt.org/z/b7hsqc1Ev

I limited it to little endian.

UsmanNadeem committed:
fix typo

Change-Id: Idaed857e4978185a74488651a5cc216c8b5eddd3

davemgreen reviewed (approved):
Thanks. LGTM

UsmanNadeem commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -20192,6 +20248,9 @@ static SDValue performIntrinsicCombine(SDNode *N,
   case Intrinsic::aarch64_sve_uqsub_x:
     return DAG.getNode(ISD::USUBSAT, SDLoc(N), N->getValueType(0),
                        N->getOperand(1), N->getOperand(2));
+  case Intrinsic::aarch64_sve_urshr:
+    return DAG.getNode(AArch64ISD::URSHR_I_PRED, SDLoc(N), N->getValueType(0),
+                       N->getOperand(1), N->getOperand(2), N->getOperand(3));
```
For naming I looked at a few other instructions that have similar behavior as `urshr`, i.e. inactive elements in the destination vector remain unmodified, and they were also named as `_PRED`.

I am quite new to the isel backend. Can you please explain what difference having a PatFrag would make compared to the code above?

paulwalker-arm replied:
Can you point to an example of where a `_PRED` node expects the results of inactive lanes to take a known value? because that really shouldn't be the case (there's a comment at the top of AArch64SVEInstrInfo.td that details the naming strategy).  The intent of the `_PRED` nodes is to allow predication to be represented at the DAG level rather than waiting until instruction selection.  They have no requirement for the results of inactive lanes to free up instruction section to allow the best use of unpredicated and/or reversed instructions.

The naming is important because people will assume the documented rules implementing DAG combines or make changes to instruction selection and thus if they're not followed it's very likely to introduce bugs.  If it's important for the ISD node to model the results of the inactive lanes in accordance with the underlying SVE instruction then it should be named as such (e.g. `URSHR_I_MERGE_OP1`).

This is generally not the case and typically at the ISD level the result of inactive lanes is not important (often because an all active predicate is passed in) and thus the `_PRED` suffix is used.  When this is the case we still want to minimise the number of ISel patterns and so a PatFrags is created to match both the ISD node and the intrinsic to the same instruction (e.g. `AArch64mla_m1`).

UsmanNadeem replied:
Thanks, got your point now. I will post a followup fix.
